### ISSUE TITLE
lock-timeout waits for all workspaces at the azure backend

### ISSUE THREAD

**Issue by StefanSchoof** [hashicorp/terraform#22242]
### Terraform Version

```
Terraform v0.12.5
+ provider.null v2.1.2
```

### Terraform Configuration Files

```hcl
terraform {
  backend "azurerm" {
    resource_group_name  = "locktest"
    storage_account_name = "locktesttfstate"
    container_name       = "tfstate"
    key                  = "terraform.tfstate"
  }
}

resource "null_resource" "wait" {
  provisioner "local-exec" {
    command = "sleep 60"
  }
}
```

### Debug Output

<details>
 <summary>Debug Output</summary>

```
2019/07/29 18:35:37 [INFO] Terraform version: 0.12.5
2019/07/29 18:35:37 [INFO] Go runtime version: go1.12.4
2019/07/29 18:35:37 [INFO] CLI args: []string{"/usr/local/bin/terraform", "plan"                                                                                                             , "-lock-timeout=2m"}
2019/07/29 18:35:37 [DEBUG] Attempting to open CLI config file: /home/name/.                                                                                                             terraformrc
2019/07/29 18:35:37 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2019/07/29 18:35:37 [INFO] CLI command args: []string{"plan", "-lock-timeout=2m"                                                                                                             }
2019/07/29 18:35:37 [TRACE] Meta.Backend: built configuration for "azurerm" back                                                                                                             end with hash value 4150761362
2019/07/29 18:35:37 [TRACE] Preserving existing state lineage "eb3f500f-157c-db4                                                                                                             7-97c5-39c76837f133"
2019/07/29 18:35:37 [TRACE] Preserving existing state lineage "eb3f500f-157c-db4                                                                                                             7-97c5-39c76837f133"
2019/07/29 18:35:37 [TRACE] Meta.Backend: working directory was previously initi                                                                                                             alized for "azurerm" backend
2019/07/29 18:35:37 [TRACE] Meta.Backend: using already-initialized, unchanged "                                                                                                             azurerm" backend configuration
2019/07/29 18:35:37 [DEBUG] Loading Environment "public"
2019/07/29 18:35:37 Testing if Service Principal / Client Certificate is applica                                                                                                             ble for Authentication..
2019/07/29 18:35:37 Testing if Service Principal / Client Secret is applicable f                                                                                                             or Authentication..
2019/07/29 18:35:37 Testing if Managed Service Identity is applicable for Authen                                                                                                             tication..
2019/07/29 18:35:37 Testing if Obtaining a token from the Azure CLI is applicabl                                                                                                             e for Authentication..
2019/07/29 18:35:37 Using Obtaining a token from the Azure CLI for Authenticatio                                                                                                             n
2019/07/29 18:35:37 [DEBUG] Resource "https://management.core.windows.net/" isn'                                                                                                             t for the correct Tenant
2019/07/29 18:35:37 [DEBUG] Resource "https://management.core.windows.net/" isn'                                                                                                             t for the correct Tenant
2019/07/29 18:35:37 [DEBUG] Resource "https://management.core.windows.net/" isn'                                                                                                             t for the correct Tenant
2019/07/29 18:35:37 [DEBUG] Resource "https://management.core.windows.net/" isn'                                                                                                             t for the correct Tenant
2019/07/29 18:35:37 [DEBUG] Resource "https://management.core.windows.net/" isn'                                                                                                             t for the correct Tenant
2019/07/29 18:35:37 [DEBUG] Resource "https://management.core.windows.net/" isn'                                                                                                             t for the correct Tenant
2019/07/29 18:35:37 [DEBUG] Resource "https://management.core.windows.net/" isn'                                                                                                             t for the correct Tenant
2019/07/29 18:35:38 [TRACE] Meta.Backend: instantiated backend of type *azure.Ba                                                                                                             ckend
2019/07/29 18:35:38 [DEBUG] checking for provider in "."
2019/07/29 18:35:38 [DEBUG] checking for provider in "/usr/local/bin"
2019/07/29 18:35:38 [DEBUG] checking for provider in ".terraform/plugins/linux_a                                                                                                             md64"
2019/07/29 18:35:38 [DEBUG] found provider "terraform-provider-null_v2.1.2_x4"
2019/07/29 18:35:38 [DEBUG] found valid plugin: "null", "2.1.2", "/home/name/locktest2/.terraform/plugins/linux_amd64/terraform-provider-null_v2.1.2_x4"
2019/07/29 18:35:38 [DEBUG] checking for provisioner in "."
2019/07/29 18:35:38 [DEBUG] checking for provisioner in "/usr/local/bin"
2019/07/29 18:35:38 [DEBUG] checking for provisioner in ".terraform/plugins/linu                                                                                                             x_amd64"
2019/07/29 18:35:38 [TRACE] Meta.Backend: backend *azure.Backend does not suppor                                                                                                             t operations, so wrapping it in a local backend
2019/07/29 18:35:38 [INFO] backend/local: starting Plan operation
2019/07/29 18:35:38 [TRACE] backend/local: requesting state manager for workspac                                                                                                             e "dev"
2019/07/29 18:35:38 [DEBUG] Building the Blob Client from an Access Token (using                                                                                                              user credentials)
2019/07/29 18:35:38 [DEBUG] Azure Backend Request:
POST /subscriptions/dd47ff80-9de3-45ec-a16d-75f3599d9f8b/resourceGroups/locktest                                                                                                             /providers/Microsoft.Storage/storageAccounts/locktesttfstate/listKeys?api-versio                                                                                                             n=2016-01-01 HTTP/1.1
Host: management.azure.com
User-Agent: Terraform/0.12.5
Content-Length: 0
Accept-Encoding: gzip


2019/07/29 18:35:38 [DEBUG] Azure Backend Response for https://management.azure.                                                                                                             com/subscriptions/dd47ff80-9de3-45ec-a16d-75f3599d9f8b/resourceGroups/locktest/p                                                                                                             roviders/Microsoft.Storage/storageAccounts/locktesttfstate/listKeys?api-version=                                                                                                             2016-01-01:
HTTP/2.0 200 OK
Cache-Control: no-cache
Content-Type: application/json
Date: Mon, 29 Jul 2019 18:35:38 GMT
Expires: -1
Pragma: no-cache
Server: Microsoft-Azure-Storage-Resource-Provider/1.0,Microsoft-HTTPAPI/2.0 Micr                                                                                                             osoft-HTTPAPI/2.0
Strict-Transport-Security: max-age=31536000; includeSubDomains
Vary: Accept-Encoding
X-Content-Type-Options: nosniff
X-Ms-Correlation-Request-Id: ddf2142d-0dec-4a49-b91b-3cc5b18d105b
X-Ms-Ratelimit-Remaining-Subscription-Writes: 1197
X-Ms-Request-Id: f0fb0881-4d76-404b-b877-b24571156c5b
X-Ms-Routing-Request-Id: WESTEUROPE:20190729T183538Z:ddf2142d-0dec-4a49-b91b-3cc                                                                                                             5b18d105b

{"keys":[{"keyName":"key1","value":"reducted","permissions":"FULL"},{"keyName":"                                                                                                             key2","value":"reducted","permissions":"FULL"}]}

Error: Error loading state: failed to lock azure state: storage: service returne                                                                                                             d error: StatusCode=409, ErrorCode=LeaseAlreadyPresent, ErrorMessage=There is al                                                                                                             ready a lease present.
RequestId:93412b6e-601e-0059-2f3c-46873b000000
Time:2019-07-29T18:35:38.4595152Z, RequestInitiated=Mon, 29 Jul 2019 18:35:37 GM                                                                                                             T, RequestId=93412b6e-601e-0059-2f3c-46873b000000, API Version=2016-05-31, Query                                                                                                             ParameterName=, QueryParameterValue=
Lock Info:
  ID:        db52235e-4604-7f5b-178e-e312ccd93dc0
  Path:      tfstate/terraform.tfstateenv:dev
  Operation: OperationTypeApply
  Who:       name@mylinuxvn
  Version:   0.12.5
  Created:   2019-07-29 18:34:52.306563841 +0000 UTC
  Info:
```
</details>

### Expected Behavior

The lock-timeout argument waits until no lock is blocking the current operation. The workspace in which the operation is does not matter. No `Error loading state: failed to lock azure state: storage: service returned error: StatusCode=409, ErrorCode=LeaseAlreadyPresent, ErrorMessage=There is already a lease present` happens

### Actual Behavior

If a operation is running in the default workspace and a plan is done in a other workspace, the second operation does not wait until the lock-timeout is over. The command is immediately ending with an error:
```
> terraform plan -lock-timeout=2m

Error: Error loading state: failed to lock azure state: storage: service returned error: StatusCode=409, ErrorCode=LeaseAlreadyPresent, ErrorMessage=There is already a lease present.
RequestId:118d8c78-201e-0067-513d-461044000000
Time:2019-07-29T18:45:00.4340748Z, RequestInitiated=Mon, 29 Jul 2019 18:45:00 GMT, RequestId=118d8c78-201e-0067-513d-461044000000, API Version=2016-05-31, QueryParameterName=, QueryParameterValue=
Lock Info:
  ID:        ce9aaf71-114e-0613-66de-13dca90827d9
  Path:      tfstate/terraform.tfstateenv:dev
  Operation: OperationTypeApply
  Who:       name@mylinuxvn
  Version:   0.12.5
  Created:   2019-07-29 18:44:51.229821367 +0000 UTC
  Info:

```

### Steps to Reproduce

1. copy terraform file into two folders (a and b)
2. adjust backend configuration
3. (a) `terraform init`
4. (b) `terraform init`
5. (a) `terraform apply -auto-approve` (do not wait until finish)
6. (b) `terraform plan -lock-timeout=2m` (works without an error)
7. (a) `terraform workspace new dev`
8. (a) `terraform destroy -auto-approve`
9. (b) `terraform workspace select dev`
10. (a) `terraform apply -auto-approve` (do not wait until finish)
11. (b) `terraform plan -lock-timeout=2m`

**Comment by teamterraform** [comment]
Hi @StefanSchoof !
I was able to reproduce the behavior you reported, but I don't think that there is a bug. 

When you run `terraform workspace new dev` in folder `a`, terraform _also_ selects workspace `dev`. Thusly from steps 7 on both folders are using the `dev` workspace and lock. 

This is the output in folder `a` after running `terraform workspace new dev`: 
```
[/tmp/scratch/a]# terraform workspace list
  default
* dev
```

So it's not unexpected that running `apply` in folder `a`(your step 1o) conflicts with `plan` in folder `b`, step `11`.

Please let me know if I've misunderstood your question and steps.

Here's the `terraform workspace new` documentation that shows that `new` also `select`s the workspace: https://www.terraform.io/docs/commands/workspace/new.html

**Comment by StefanSchoof** [comment]
Hi @teamterraform ,

thanks for the reply. I do not think that the workspace of the (a) operation matters. If I start a terraform command with a `-lock-timeout=2m` I expect the command to not fail immediately with a locking issue. If there is a lock for my operation the command should wait. If there is no lock for this operation the operation should succeed. (Same behavior like in step 6 of the repro) I agree that there should be a conflict, but the timeout should handle that.

I run the terraform commands in a CI system and this behavior breaks builds. Running a other build should should wait until terraform can get a lock.

**Comment by teamterraform** [comment]
I think this is the key statement from your last comment:
> If there is a lock for my operation the command should wait.

The [state lock documentation](https://www.terraform.io/docs/state/locking.html)  explicitly states that terraform will _not_ wait when state is locked. 

I can re-label this issue as a feature request, if you'd like.

**Comment by StefanSchoof** [comment]
But the documentation for lock-timeout describe this as "Duration to retry a state lock." https://www.terraform.io/docs/commands/plan.html#lock-timeout-0s and the default value of 0s is exactly the behavior described in https://www.terraform.io/docs/state/locking.html

And without a workspace this is exactly working as I expect. I think if this is the desired behavior the documentation for the `lock-timeout` argument should explain the restrictions under which it is working.

**Comment by teamterraform** [comment]
Ah, I see! Thank you for pointing that out. I will re-label this issue.

**Comment by StefanSchoof** [comment]
You removed the backend/azurerm tag. I think there is good possibility this is a bug in the azurerm backend. I would be surprised if the more used aws s3 backend has the same behavior and nobody opened a issue. (I have currently no easy way to test that.)

**Comment by tareks** [comment]
Any updates on this? It seems like this is still an issue with 0.12.26. 

Even when supplying a `-lock-timeout=300s` seems to result in the following error:
```
Failed to load state: failed to lock azure state: storage: service returned error: StatusCode=409, ErrorCode=LeaseAlreadyPresent, ErrorMessage=There is already a lease present.
```

**Comment by wawe** [comment]
I also ran into this issue with v0.14.8, but could not reproduce it anymore with v0.15.0-beta1. I used `git bisect` to track down a likely fix in 85b9bdea9626791186e58576d4dd509dfc7d650a.

**Comment by mbfrahry** [comment]
Thanks for checking this @wawe! I'll go ahead and close this down and it'll be released with Terraform 0.15! Fixed with #26561
